### PR TITLE
nixos/cloud-init: fix hostname and resolvconf configuration

### LINKED ISSUE #195966: cloud-init: need update to 22.3.4
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: cloud-init

current version: 21.4
desired version: 22.3.4

###### Notify maintainers

maintainers: @madjar @phile314

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by illustris**
Closes: #195966

###### Description of changes

- Fix hostname configuration on proxmox, which uses "hostname" in user-data instead of "local-hostname" in meta-data.
- Allow setting resolv.conf through cloud-init
- remove update_etc_hosts, as /etc/hosts is managed by nixos
- Add tests for new changes
- Add timeouts to make tests fail faster

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/virtualization/cloud-init/default.nix [hunk 1]
```diff
@@ -96,19 +103,7 @@ python3.pkgs.buildPythonApplication rec {
...
-    "tests/unittests/test_datasource/test_aliyun.py"
-    "tests/unittests/test_datasource/test_azure.py"
-    "tests/unittests/test_datasource/test_ec2.py"
-    "tests/unittests/test_datasource/test_exoscale.py"
-    "tests/unittests/test_datasource/test_gce.py"
-    "tests/unittests/test_datasource/test_openstack.py"
-    "tests/unittests/test_datasource/test_scaleway.py"
-    "tests/unittests/test_ec2_util.py"
-  ];
+  disabledTestPaths = [];
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/cloud-init/default.nix:107]:
```suggestion
```

**Comment by illustris** [comment]
https://github.com/NixOS/nixpkgs/pull/195505 marked the test as broken. This PR should also prevent cases where a failing test runs till the test times out.

File: nixos/tests/cloud-init.nix [hunk 2]
```diff
@@ -49,19 +49,17 @@ let
               gateway: '12.34.56.9'
           - type: nameserver
             address:
-            - '8.8.8.8'
+            - '6.7.8.9'
```
Comment (Mindavi) [inline nixos/tests/cloud-init.nix:52]: Why this change? It's not clear to me what the difference is

Comment (illustris) [reply nixos/tests/cloud-init.nix:52]: Because systemd-resolved automatically adds cloudflare DNS (1.1.1.1) and google DNS (8.8.8.8) as fallback. I wanted to be sure that it is picking up the nameserver from cloud-init instead of using the fallback values.

Comment (Mindavi) [reply nixos/tests/cloud-init.nix:52]: Might be good to document that

Comment (illustris) [reply nixos/tests/cloud-init.nix:52]: Sure, I'll add a comment in the next PR involving cloud-init
